examples/advanced/ocus_explanations.py

```diff
@@ -0,0 +1,302 @@
+
+    ## OPT Model: Variables
+    hs_vars = boolvar(shape=len(Iend))
+    F = I | set(~var for var in Iend - I)
+
+    # id of variables that need to be explained
+    remaining_hs_vars = hs_vars[[id for id, var in enumerate(F) if var not in I]]
+
+    # mapping between hitting set variables hs_var <-> Iend
+    hs_vars_to_Iend = dict( (hs_vars[id], var) for id, var in enumerate(F))
+    Iend_to_hs_vars = dict( (var, hs_vars[id]) for id, var in enumerate(F))
+
+
+    # CONDITIONAL OPTIMISATION MODEL
+    hs_model = Model(
+        # exactly one variable to explain!
+        sum(remaining_hs_vars) == 1,
+        # optimal hitting set
+        minimize=sum(hs_var * cost(hs_vars_to_Iend[hs_var]) for hs_var in hs_vars)
+    )
+
+    ## instantiate hitting set solver
+    hittingset_solver = SolverLookup.lookup(solver)(hs_model)
+
+    while(True):
+        hittingset_solver.solve()
+
+        # Get hitting set
+        hs = hs_vars[hs_vars.value() == 1]
+
+        # map to vars of formula F
+        S = set(hs_vars_to_Iend[hs_var] for hs_var in hs)
+
+        if verbose > 0:
+            print("\n\t hs =", hs, S)
+
+        # SAT check and computation of model
+        if not sat.solve(assumptions=S):
+            if verbose > 0:
+                print("\n\t ===> OCUS =", S)
+
+            # deleting the hitting set solver
+            return S
+        # satisfying model
+        S = set(v == v.value() for v in F)
+
+        # compute complement of model in formula F
+        C =  F - S
+        set_to_hit = set(Iend_to_hs_vars[Iend_var] for Iend_var in F - S)
+
+        # Add complement as a new set to hit: sum x[j] * hij >= 1
+        hittingset_solver += (sum(set_to_hit) >= 1)
+
+        if verbose > 0:
+            print("\t Complement =", C)
+            print("\t set-to-hit =", set_to_hit)
+
+def optimal_propagate(hard, soft, solver="ortools", verbose=False):
+    """
+        optimal_propagate produces the intersection of all models of cnf more precise
+        projected on focus.
+
+        Args:
+        hard (list): List of hard constraints
+
+        soft (iterable): List of soft constraints
+
+        I (set): partial model of sat literals
+
+        user_vars (list):
+            +/- selected boolean variables of the sat solver
+    """
+    # Build sat model
+    sat_model = Model(hard)
+    sat = SolverLookup.lookup(solver)(sat_model)
+    assert sat.solve(assumptions=soft), "Propagation of soft constraints only possible if model is SAT."
+
+    # Extracting only relevant variables
+    user_vars = get_variables_model(sat_model)
+
+    # initial model that needs to be refined
+    sat_model = set(v == v.value() for v in user_vars)
+
+    if verbose > 1:
+        print("\nOptimal Propagate")
+        print("-" * len("Optimal Propagate"))
+        print("\nInitial sat model:", sat_model)
+
+    while(True):
+        # negate the values of the model
+        blocking_clause = ~all(v == v.value() for v in sat_model)
```

Comment:
line 276 says that sat_model is itself a set of (v == v.value()) constraints, then v in sat_model then be a variable?